File: dz9/memc_load_multi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import gzip
import sys
import glob
import logging
import collections
from multiprocessing import Pool, cpu_count
from optparse import OptionParser
from queue import Queue, Empty
from threading import Thread

import appsinstalled_pb2
import memcache

NORMAL_ERR_RATE = 0.01
AppsInstalled = collections.namedtuple("AppsInstalled", ["dev_type", "dev_id", "lat", "lon", "apps"])
FINISH = 'finish'


def dot_rename(path):
    print('path============', path)
    head, fn = os.path.split(path)
    # atomic in most cases
    # os.rename(path, os.path.join(head, "." + fn))


def insert_appsinstalled(memc, appsinstalled, dry_run=False):
    ua = appsinstalled_pb2.UserApps()
    ua.lat = appsinstalled.lat
    ua.lon = appsinstalled.lon
    key = "%s:%s" % (appsinstalled.dev_type, appsinstalled.dev_id)
    ua.apps.extend(appsinstalled.apps)
    packed = ua.SerializeToString()
    # @TODO persistent connection
    # @TODO retry and timeouts!
    try:
        if dry_run:
            logging.debug("%s - %s -> %s" % (memc, key, str(ua).replace("\n", " ")))
        else:
            # memc = memcache.Client([memc_addr])
            memc.set(key, packed)
    except Exception as e:
        logging.exception("Cannot write to memc %s: %s" % (memc, e))
        return False
    return True


def parse_appsinstalled(line):
    line = str(line)
    line_parts = line.strip().split("\t")
    if len(line_parts) < 5:
        return
    dev_type, dev_id, lat, lon, raw_apps = line_parts
    if not dev_type or not dev_id:
        return
    try:
        apps = [int(a.strip()) for a in raw_apps.split(",")]
    except ValueError:
        apps = [int(a.strip()) for a in raw_apps.split(",") if a.isidigit()]
        logging.info("Not all user apps are digits: `%s`" % line)
    try:
        lat, lon = float(lat), float(lon)
    except ValueError:
        logging.info("Invalid geo coords: `%s`" % line)
    return AppsInstalled(dev_type, dev_id, lat, lon, apps)


class Worker(Thread):
    def __init__(self, job_pool, memc, results_queue):
        super().__init__()
        self.job_pool = job_pool
        self.memc = memc

    def run(self) -> None:
        processed = 0
        errors = 0
        while True:
            try:
                line = self.job_pool.get(timeout=0.1)
                print('line====', line)
                if line == FINISH:
                    self.job_pool.task_done()
                    break
                else:
                    processed += 1
                    if not insert_appsinstalled(self.memc, line):
                        errors += 1

                    # appsinstalled = parse_appsinstalled(line)
                    # print('appsinstalled====', appsinstalled)
                    # insert_appsinstalled(self.memc, appsinstalled)
                    # proc_items, err_items = self.insert_appsinstalled(chunks)
                    # print('self.memc=============', self.memc)
                    # proc_items, err_items = insert_appsinstalled(self.memc, line)
                    # processed += proc_items
                    # errors += err_items
                    self.job_pool.task_done()
            except Empty:
                continue


def file_handler(fn, options, device_memc):
    # print(fn, device_memc)
    jobs_pool = {}
    treads_pool = list()
    results_queue = Queue()
    for dev_type, address in device_memc.items():
        # print(dev_type, address)
        memc = memcache.Client(servers=[address])
        jobs_pool[dev_type] = Queue()
        thread = Worker(jobs_pool[dev_type], memc, results_queue)
        treads_pool.append(thread)
        thread.start()

    processed = errors = 0
    logging.info('Processing %s' % fn)
    fd = gzip.open(fn)

    for line in fd:
        line = line.decode().strip()
        if not line:
            continue
        print('line2=', line)
        appsinstalled = parse_appsinstalled(line)
        print('appsinstalled==================', appsinstalled)
        jobs_pool[appsinstalled.dev_type].put(appsinstalled)

        # if not appsinstalled:
        #     errors += 1
        #     continue
        # memc_addr = device_memc.get(appsinstalled.dev_type)
        # if not memc_addr:
        #     errors += 1
        #     logging.error("Unknown device type: %s" % appsinstalled.dev_type)
        #     continue

    for d_type in device_memc:
        jobs_pool[d_type].put(FINISH)

    for tp in treads_pool:
        tp.join()

    if not processed:
        fd.close()
        # continue

    # err_rate = float(errors) / processed
    # if err_rate < NORMAL_ERR_RATE:
    #     logging.info("Acceptable error rate (%s). Successfull load" % err_rate)
    # else:
    #     logging.error("High error rate (%s > %s). Failed load" % (err_rate, NORMAL_ERR_RATE))
    fd.close()
    # dot_rename(fn)
    return fn


def main(options):
    device_memc = {
        "idfa": options.idfa,
        "gaid": options.gaid,
        "adid": options.adid,
        "dvid": options.dvid,
    }

    pool = Pool(cpu_count())

    process_args = ((fn, options, device_memc) for fn in sorted(glob.iglob(options.pattern)))
    for file_name in pool.starmap(file_handler, process_args):
        dot_rename(file_name)


def prototest():
    sample = "idfa\t1rfw452y52g2gq4g\t55.55\t42.42\t1423,43,567,3,7,23\ngaid\t7rfw452y52g2gq4g\t55.55\t42.42\t7423,424"
    for line in sample.splitlines():
        dev_type, dev_id, lat, lon, raw_apps = line.strip().split("\t")
        apps = [int(a) for a in raw_apps.split(",") if a.isdigit()]
        lat, lon = float(lat), float(lon)
        ua = appsinstalled_pb2.UserApps()
        ua.lat = lat
        ua.lon = lon
        ua.apps.extend(apps)
        packed = ua.SerializeToString()
        unpacked = appsinstalled_pb2.UserApps()
        unpacked.ParseFromString(packed)
        assert ua == unpacked


if __name__ == '__main__':
    op = OptionParser()
    op.add_option("-t", "--test", action="store_true", default=False)
    op.add_option("-l", "--log", action="store", default=None)
    op.add_option("--dry", action="store_true", default=False)
    op.add_option("--pattern", action="store", default="/data/appsinstalled/*.tsv.gz")
    op.add_option("--idfa", action="store", default="127.0.0.1:33013")
    op.add_option("--gaid", action="store", default="127.0.0.1:33014")
    op.add_option("--adid", action="store", default="127.0.0.1:33015")
    op.add_option("--dvid", action="store", default="127.0.0.1:33016")
    (opts, args) = op.parse_args()
    logging.basicConfig(filename=opts.log, level=logging.INFO if not opts.dry else logging.DEBUG,
                        format='[%(asctime)s] %(levelname).1s %(message)s', datefmt='%Y.%m.%d %H:%M:%S')
    if opts.test:
        prototest()
        sys.exit(0)

    logging.info("Memc loader started with options: %s" % opts)
    try:
        main(opts)
    except Exception as e:
        logging.exception("Unexpected error: %s" % e)
        sys.exit(1)

File: dz12/memc_load_multi.go
package main

import (
	"./appsinstalled"

	"bufio"
	"compress/gzip"
	"errors"
	"flag"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"

	"github.com/bradfitz/gomemcache/memcache"
	"github.com/golang/protobuf/proto"
)

type AppsInstalled struct {
	devType string
	devId   string
	lat     float64
	lon     float64
	apps    []uint32
}

type Options struct {
	logfile string
	workers int
	buffer  int
	pattern string
}

type MemcacheItem struct {
	key  string
	data []byte
}

type Result struct {
	processed int
	errors    int
}

const NORMAL_ERR_RATE = 0.01

func createMemcaches(options *Options, resultsQueue chan Result) map[string]chan *MemcacheItem {

	memcQueues := make(map[string]chan *MemcacheItem)

	deviceMemc := map[string]string{
		"idfa": idfa,
		"gaid": gaid,
		"adid": adid,
		"dvid": dvid,
	}

	for devType, memcAddr := range deviceMemc {
		memcQueues[devType] = make(chan *MemcacheItem, options.buffer)
		m := memcache.New(memcAddr)
		go MemcacheWorker(m, memcQueues[devType], resultsQueue)
	}

	return memcQueues
}

func MemcacheWorker(mc *memcache.Client, items chan *MemcacheItem, resultsQueue chan Result) {
	processed, errs := 0, 0

	for {
		task, ok := <-items
		if !ok {
			resultsQueue <- Result{processed, errs}
			log.Printf("memcache processed %d | errors %d", processed, errs)
			return
		}

		err := mc.Set(&memcache.Item{
			Key:   task.key,
			Value: task.data,
		})

		if err != nil {
			errs++
			log.Println(err)
		} else {
			processed++
		}
	}
}

func LineParser(lines chan string, memcQueues map[string]chan *MemcacheItem, resultsQueue chan Result) {
	errs := 0
	for line := range lines {
		appsInstalled, err := parseAppsInstalled(line)
		if err != nil {
			errs += 1
			continue
		}
		item, err := SerializeAppsInstalled(appsInstalled)
		if err != nil {
			errs += 1
			continue
		}
		queue, ok := memcQueues[appsInstalled.devType]
		if !ok {
			log.Println("Unknown device type:", appsInstalled.devType)
			errs += 1
			continue
		}
		queue <- item
	}
	resultsQueue <- Result{errors: errs}
}

func parseAppsInstalled(line string) (*AppsInstalled, error) {
	line = strings.Trim(line, "")
	unpacked := strings.Split(line, "\t")
	if len(unpacked) != 5 {
		return nil, errors.New("Bad line: " + line)
	}
	devType, devId, rawLat, rawLon, rawApps := unpacked[0], unpacked[1], unpacked[2], unpacked[3], unpacked[4]
	lat, err := strconv.ParseFloat(rawLat, 64)
	if err != nil {
		return nil, err
	}
	lon, err := strconv.ParseFloat(rawLon, 64)
	if err != nil {
		return nil, err
	}
	var apps []uint32
	for _, rawApp := range strings.Split(rawApps, ",") {
		if app, err := strconv.ParseUint(rawApp, 10, 32); err == nil {
			apps = append(apps, uint32(app))
		}
	}

	return &AppsInstalled{
		devType: devType,
		devId:   devId,
		lat:     lat,
		lon:     lon,
		apps:    apps,
	}, nil
}

func SerializeAppsInstalled(appsInstalled *AppsInstalled) (*MemcacheItem, error) {
	aua := &appsinstalled.UserApps{
		Lat:  proto.Float64(appsInstalled.lat),
		Lon:  proto.Float64(appsInstalled.lon),
		Apps: appsInstalled.apps,
	}
	key := fmt.Sprintf("%s:%s", appsInstalled.devType, appsInstalled.devId)
	packed, err := proto.Marshal(aua)
	if err != nil {
		return nil, err
	}
	return &MemcacheItem{key, packed}, nil
}

func fileReader(filename string, linesQueue chan string) error {
	log.Println("Processing:", filename)
	f, err := os.Open(filename)
	if err != nil {
		log.Printf("Can't open file: %s", filename)
		return err
	}
	defer f.Close()

	gz, err := gzip.NewReader(f)
	if err != nil {
		log.Printf("Can't create Reader %v", err)
		return err
	}
	defer gz.Close()

	scanner := bufio.NewScanner(gz)
	for scanner.Scan() {
		line := scanner.Text()
		line = strings.Trim(line, " ")
		if line == "" {
			continue
		}
		linesQueue <- line
	}

	if err := scanner.Err(); err != nil {
		log.Printf("Scanner error: %v", err)
		return err
	}

	return nil
}

func dotRename(path string) error {
	dir, f := filepath.Split(path)
	if err := os.Rename(path, dir+"."+f); err != nil {
		log.Printf("Can't rename a file: %s", path)
		return err
	}
	return nil
}

func mainProcess(options *Options) error {
	files, err := filepath.Glob(options.pattern)
	if err != nil {
		log.Printf("Files not found for pattern: %s", options.pattern)
		return err
	}
	resultsQueue := make(chan Result)
	memcQueues := createMemcaches(options, resultsQueue)

	deviceMemc := map[string]string{
		"idfa": idfa,
		"gaid": gaid,
		"adid": adid,
		"dvid": dvid,
	}

	for devType, memcAddr := range deviceMemc {
		memcQueues[devType] = make(chan *MemcacheItem, options.buffer)
		mc := memcache.New(memcAddr)
		go MemcacheWorker(mc, memcQueues[devType], resultsQueue)
	}

	linesQueue := make(chan string, options.buffer)
	for i := 0; i < options.workers; i++ {
		go LineParser(linesQueue, memcQueues, resultsQueue)
	}

	sort.Strings(files)
	for _, filename := range files {
		fileReader(filename, linesQueue)
		//dotRename(filename)
	}
	close(linesQueue)

	processed, errs := 0, 0
	for i := 0; i < options.workers; i++ {
		results := <-resultsQueue
		processed += results.processed
		errs += results.errors
	}

	for _, queue := range memcQueues {
		close(queue)
		results := <-resultsQueue
		processed += results.processed
		errs += results.errors

	}

	errRate := float32(errs) / float32(processed)
	if errRate < NORMAL_ERR_RATE {
		log.Printf("Acceptable error rate (%g). Successfull load\n", errRate)
	} else {
		log.Printf("High error rate (%g > %g). Failed load\n", errRate, NORMAL_ERR_RATE)
	}

	return nil
}

var (
	logfile string
	pattern string
	workers int
	buffer  int
	idfa    string
	gaid    string
	adid    string
	dvid    string
)

func init() {
	flag.StringVar(&logfile, "log", "", "")
	flag.StringVar(&pattern, "pattern", "/data/*.tsv.gz", "")
	flag.IntVar(&workers, "workers", 4, "")
	flag.IntVar(&buffer, "buffer", 100, "")
	flag.StringVar(&idfa, "idfa", "127.0.0.1:33013", "")
	flag.StringVar(&gaid, "gaid", "127.0.0.1:33014", "")
	flag.StringVar(&adid, "adid", "127.0.0.1:33015", "")
	flag.StringVar(&dvid, "dvid", "127.0.0.1:33016", "")
}

func main() {
	flag.Parse()

	options := &Options{
		logfile: logfile,
		pattern: pattern,
		workers: workers,
		buffer:  buffer,
	}

	if options.logfile != "" {
		f, err := os.OpenFile(options.logfile, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
		if err != nil {
			log.Fatalf("Cannot open log file: %s", options.logfile)
		}
		defer f.Close()
		log.SetOutput(f)
	}
	log.Println("Memc loader started with options: ", *options)
	mainProcess(options)
}
